core/src/main/scala/org/apache/spark/resource/ExecutorResourceRequest.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.resource
+
+/**
+ * An Executor resource request. This is used in conjunction with the ResourceProfile to
+ * programmatically specify the resources needed for an RDD that will be applied at the
+ * stage level.
+ *
+ * This is used to specify what the resource requirements are for an Executor and how
+ * Spark can find out specific details about those resources. Not all the parameters are
+ * required for every resource type. The resources names supported
+ * correspond to the regular Spark configs with the prefix removed. For instance overhead
+ * memory in this api is memoryOverhead, which is spark.executor.memoryOverhead with
+ * spark.executor removed. Resources like GPUs are resource.gpu
+ * (spark configs spark.executor.resource.gpu.*). The amount, discoveryScript, and vendor
+ * parameters for resources are all the same parameters a user would specify through the
+ * configs: spark.executor.resource.{resourceName}.{amount, discoveryScript, vendor}.
+ *
+ * For instance, a user wants to allocate an Executor with GPU resources on YARN. The user has
+ * to specify the resource name (resource.gpu), the amount or number of GPUs per Executor,
+ * units would not be used as its not a memory config, the discovery script would be specified
+ * so that when the Executor starts up it can discovery what GPU addresses are available for it to
+ * use because YARN doesn't tell Spark that, then vendor would not be used because
+ * its specific for Kubernetes.
+ *
+ * See the configuration and cluster specific docs for more details.
+ *
+ * There are alternative constructors for working with Java.
+ *
+ * @param resourceName Name of the resource
+ * @param amount Amount requesting
+ * @param units Optional units of the amount. For things like Memory, default is no units, only byte
+ *              types (b, mb, gb, etc) are currently supported.
```

Comment:
I think the use of units here is really confusing -- you'd need to add in more code later to support comparison of units, etc.

if you really just want to make it easy to specify units for memory, how about adding a util constructor or factory method just for memory?  `ExecutorResourceRequest.memory(amount: double, units: String)` or `ExecutorResourceRequest.memoryGB(amount: Double)`?

Reply:
yeah the code that needs to use the units isn't in this pr and I agree that could get ugly to compare units. This started out just being modeled after the YARN api. How complex that gets somewhat depends on if when comparing ResourceProfiles, the units parameter has to match.  For instance if I specify memory as 1024m and then I compare to a profile with 1g, do we consider that the same?  If we don't care about that we can just normalize the memory to be bytes or MB internally.  Alternatively we could get rid of the units and just say memory amount must be specified in MB (similar to your suggestion ExecutorResourceRequest.memoryGB(amount: Double)) , but that doesn't match the way we do the other memory configs.

I'm not sure how I feel about the api being through the constructor for everything except for memory. I can see that being confusing to the user as well.
We could remove all the parameters from the constructor and make setters for all the things we support - memory, overheadMemory, core, resources (like GPUs), etc.  This would remove the need for the Enum/String checks to see if we support it.  Really at that point you could put directly in ResourceProfile and not need ExecutorResourceRequest.

If we don't like any of those options, I could remove the units parameter for now from this PR and can add it back in with the PR that uses it.  Which will be the scheduler code for merging and scheduling tasks and the cluster manager code.
thoughts?

Reply:
I like having a list of specific methods for the types, personally I think that is clearer (you get code completion to tell you the options, rather than having to look through docs).  I guess I was wondering if you specifically wanted more extensibility here by having looser typing, eg. for compatibility.  But I don't see a compelling case for that.  You've got a hard check that the type is understood;  and its not like you could take code written for some future version of spark with another resource type, and then run it against an old version of spark, as your resource requirement wouldn't be honored on the old spark.

also I don't think I was super clear about that extra method for memory -- I meant that you could still use the general constructor, but it would only be for bytes.  Then those other methods would just be conveniences.  They would return something in bytes, and all comparisons would then be done in bytes.

Reply:
Ok the api taking bytes by default makes more sense.
Let me look at changing to just have setters directly to see what that looks like.  I was originally thinking making it more generic would make it easier to add things in the future.  The api would change just now pass in new things, bit not sure that really matters. I do envision people wanting many more options here like the dynamic allocation min max, potentially many other spark confs so need to look at how many apis that adds

Reply:
so I'll write it out a bit more but I think I like the setters for each, the only thing is whether they are in the ResourceProfile class or you keep them in the ExecutorResourceRequest type class. The reason to keep them out of ResourceProfile is to allow it to more easily handle future things like having a ResourceProfile.prefer - which would have preference of a Resource but not a strict requirement.  It also makes is so you have a bit cleaner separation of Executor and Task resources.
for example the api could be:

val rp = new ResourceProfile()
rp.executorMemory(1, "g")
rp.executorCores(2)
rp.executorResources("gpu", 1, "discoveryScript:)
rp.taskResources("gpu", 1)

Or keep it like:
val rp = new ResourceProfile()
val e = new ExecutorResourceRequest()
e.memory(1, "g")
e.cores(2)
e.resources("gpu", 1, "discoveryScript")
val t = new TaskResourceRequest()
t.resources("gpu", 1, "discoveryScript")
rp.require(e)
rp.require(t)

The first one is shorter but the second one gives us more flexibility.
 Alternatively for things like ResourceProfile.prefer would could come up with other ways to support it like having optional parameter to the api. 

thoughts on that?

Reply:
I quickly ran into issues with the second option I talked about above because the TaskResourceRequest and ExecutorResourceRequest then become mutable which I think complicates things.

I came up with a compromise that I think still gives us ability to add things like .prefer to ResourceProfiles and that is just adding an intermediate class TAskResourceRequests and ExecutorResourceRequests (note the s on the end). In these classes it adds the convenient setters for cpu, memory, etc. But it also leaves the base ExecutorResourceRequest and TaskResourceRequest classes.  I was going to make those internal classes but the downside to that is if we want the user to be able to get the resources back from the ExecutorResourceRequests those classes are essentially needed.

Here is an example:

With the existing PR:
```
-    val cpuTaskReq = new TaskResourceRequest(ResourceProfile.CPUS, 1)
-    val coresExecReq = new ExecutorResourceRequest(ResourceProfile.CORES, 2)
-    val memExecReq = new ExecutorResourceRequest(ResourceProfile.MEMORY, 4096, "mb")
-    val omemExecReq = new ExecutorResourceRequest(ResourceProfile.OVERHEAD_MEM, 2048)
-    val pysparkMemExecReq = new ExecutorResourceRequest(ResourceProfile.PYSPARK_MEM, 1024)
-
-    rprof.require(cpuTaskReq)
-    rprof.require(coresExecReq)
-    rprof.require(memExecReq)
-    rprof.require(omemExecReq)
-    rprof.require(pysparkMemExecReq)
```

With the changes I talked about adding in ExecutorResourceRequests and TaskResourceRequests:
```
+    val ereqs = new ExecutorResourceRequests()
+    ereqs.cores(2).memory(4096, "m")
+    ereqs.memoryOverhead(2048, "m").pysparkMemory(1024, "m")
+    val treqs = new TaskResourceRequests()
+    treqs.cpus(1)
+
+    rprof.require(treqs)
+    rprof.require(ereqs)
 
```

This leaves us the flexibility in the future to add things like .prefer to ResourceProfile.
thoughts on this?  I'll probably push these changes shortly for you to take a look at the code. I can always role back if we don't like it.

Reply:
yes I do like that version.

So, I realized I didn't understand your initial version of the code, though -- I thought you had a hard requirement that the resources were in the set of things you had listed.  I had missed that check `|| rName.startsWith(RESOURCE_DOT)`.  The one advantage of leaving it as strings is that you get some extensibility for free.  Eg., with the old code, would things just work if the user requested FPGA or quantum cpus, if the resource discovery scripts returned them?  both of those sound a little far-fetched to me, so I'm OK requiring changes to code before supporting new things, but just wnated to discuss it.  I think I'd still prefer static methods.

Reply:
So either the new code or the old code both support extensibility for custom resources (gpu, fpga, etc). That doesn't change with either approach.  You do that via the ExecutorResourceRequests.resource(...)  call.  I guess I didn't put an example of that in my previous comment.  I realize resource() is potentially a bit vague so thought about renaming it - maybe customResource().  But the configs for those resources are spark.executor.resource.{resourceName} (spark.executor.resource.gpu, spark.executor.resource.fpga, etc).  If you have thoughts on better name let me know.

So the only thing that is really hardcoded are those resources Spark already strictly handles - heap memory, overhead memory, cores, pyspark memory, etc... Any other custom resource that one could allocate (say from YARN or k8s) you can still specify here via that resource() option: https://github.com/apache/spark/pull/26284/files#diff-f251b8505059be76856f40dae5f8f8a0R141

I wouldn't want to remove that functionality.

Reply:
enhanced version of my example using GPU resource:

```
val ereqs = new ExecutorResourceRequests()
ereqs.cores(2).memory(4096, "m")
ereqs.memoryOverhead(2048, "m").pysparkMemory(1024, "m")
**ereqs.resource("resource.gpu", 2, "discoveryScript")**
val treqs = new TaskResourceRequests()
treqs.cpus(1)
**treqs.resource("resource.gpu", 1.0)**

rprof.require(treqs)
rprof.require(ereqs)
```

Reply:
ok, I like this new api, thanks for clarifying.

I think it might make sense to add a static method for gpu, too -- imo that actually seems like the most common use case for this entire api -- but that can be added later.